lib/FusionInventory/Agent/Task/Inventory/Linux/Storages.pm

```diff
@@ -37,15 +37,39 @@ sub doInventory {
 sub _getDevices {
     my (%params) = @_;
 
-    my $root   = $params{root};
+    my $root = $params{root};
 
     my %sources = (
-        'DESCRIPTION'  => [ \&_getHdparmInfo, \&getInfoFromSmartctl ],
-        'DISKSIZE'     => [ \&_getHdparmInfo, \&getInfoFromSmartctl ],
-        'FIRMWARE'     => [ \&_getHdparmInfo, \&getInfoFromSmartctl ],
-        'MANUFACTURER' => [ \&getInfoFromSmartctl ],
-        'MODEL'        => [ \&_getHdparmInfo, \&getInfoFromSmartctl ],
-        'WWN'          => [ \&_getHdparmInfo ],
+        'DESCRIPTION' => {
+            subs    => [ \&getInfoFromSmartctl, \&_getHdparmInfo ],
+            next_if => sub { 1 },
+        },
+        'DISKSIZE' => {
+            subs    => [ \&getInfoFromSmartctl, \&_getHdparmInfo ],
+            next_if => sub { 1 },
+        },
+        'FIRMWARE' => {
+            subs    => [ \&getInfoFromSmartctl, \&_getHdparmInfo ],
+            next_if => sub { 1 },
+        },
+        'INTERFACE' => {
+            subs    => [ \&getInfoFromSmartctl, \&_getHdparmInfo ],
+            next_if => sub { 1 },
+        },
```

Comment:
This is finally a lot a redundant definitions (also with l.69-72). I would prefer you change the algorithm at l. 76-78 to still include `[ \&getInfoFromSmartctl, \&_getHdparmInfo ]` as default subs and add few line to handle MANUFACTURER & MODEL cases. I even don't like the `sub { 0 }` & `sub { 1 }` tricks just to simply the l.101 which is finally hard to read without comment.

Reply:
In my opinion it's easier to comprehend a dispatch table than a tree of IFs.